diff --git a/judge/judgedaemon.main.php b/judge/judgedaemon.main.php
index 262d3a100..83178fcc6 100644
--- a/judge/judgedaemon.main.php
+++ b/judge/judgedaemon.main.php
@@ -1410,6 +1410,15 @@ function judge(array $judgeTask): bool
         $passdir = $testcasedir . '/' . $passCnt;
         mkdir($passdir, 0755, true);
 
+        if ($passCnt > 1) {
+            $cpcmd = 'cp -R ' . $passdir . '/../' . ($passCnt - 1) . '/feedback ' . $passdir . '/';
+            system($cpcmd);
+            logmsg(LOG_INFO, "    Copying feedback dir from pass " . ($passCnt - 1) . ': ' . $cpcmd);
+            $rmcmd = 'rm ' . $passdir . '/feedback/nextpass.in';
+            system($rmcmd);
+            logmsg(LOG_INFO, "    Executing " . $rmcmd);
+        }
+
         // Copy program with all possible additional files to testcase
         // dir. Use hardlinks to preserve space with big executables.
         $programdir = $passdir . '/execdir';
diff --git a/judge/testcase_run.sh b/judge/testcase_run.sh
index 7cffaa7ec..fa316ee76 100755
--- a/judge/testcase_run.sh
+++ b/judge/testcase_run.sh
@@ -195,7 +195,7 @@ if [ $COMBINED_RUN_COMPARE -eq 1 ]; then
 	# A combined run and compare script may now already need the
 	# feedback directory, and perhaps access to the test answers (but
 	# only the original that lives outside the chroot).
-	mkdir feedback
+	mkdir -p feedback
 	RUNARGS="$RUNARGS $TESTOUT compare.meta feedback"
 fi
 
@@ -232,8 +232,8 @@ if [ $COMBINED_RUN_COMPARE -eq 0 ]; then
 
 	exitcode=0
 	# Create dir for feedback files and make it writable for $RUNUSER
-	mkdir feedback
-	chmod a+w feedback
+	mkdir -p feedback
+	chmod -R a+w feedback
 
 	runcheck $GAINROOT "$RUNGUARD" ${DEBUG:+-v} $CPUSET_OPT -u "$RUNUSER" -g "$RUNGROUP" \
 		-m $SCRIPTMEMLIMIT -t $SCRIPTTIMELIMIT --no-core \
